core/vm/eips.go

```diff
@@ -1055,13 +1060,20 @@ func opExtCall(pc *uint64, interpreter *EVMInterpreter, scope *ScopeContext) ([]
 	if interpreter.readOnly && !value.IsZero() {
 		return nil, ErrWriteProtection
 	}
-	if !value.IsZero() {
-		gas += params.CallStipend
-	}
 
-	ret, returnGas, err := interpreter.evm.Call(scope.Contract, toAddr, args, gas, &value)
+	var (
+		ret       []byte
+		returnGas uint64
+		err       error
+	)
+	if interpreter.evm.callGasTemp == 0 {
+		// zero temp call gas indicates a min retained gas error
+		ret, returnGas, err = nil, 0, ErrExecutionReverted
+	} else {
+		ret, returnGas, err = interpreter.evm.Call(scope.Contract, toAddr, args, gas, &value)
+	}
 
-	if err == ErrExecutionReverted {
+	if errors.Is(err, ErrExecutionReverted) || errors.Is(err, ErrInsufficientBalance) || errors.Is(err, ErrDepth) {
```

Comment:
I suspect that the error conditions here, and in general having error-introspection, is going to be very problematic going forward. 
We've traditionally been pretty hand-wavy about errors, everything is more or less "out of gas error", the exception being revert which doesn't eat all gas. But now all implementations will all of a sudden have to take great care about specific individual errors. 

I think that is a mistake. But not really related to this specific PR

Reply:
The design intent was to separate logic caused the failure (code 2) from the vm caused the failure (code 1)